transport/handler_server.go

```diff
@@ -319,7 +320,16 @@ func (ht *serverHandlerTransport) HandleStreams(startStream func(*Stream)) {
 				s.buf.put(&recvMsg{data: buf[:n]})
 			}
 			if err != nil {
-				s.buf.put(&recvMsg{err: err})
+				if err, ok := err.(http2.StreamError); ok {
+					code := http2ErrConvTab[err.Code]
+					if code == codes.OK {
+						s.buf.put(&recvMsg{err: io.EOF})
+					} else {
+						s.buf.put(&recvMsg{err: StreamErrorf(code, "%v", err)})
+					}
+				} else {
+					s.buf.put(&recvMsg{err: io.EOF})
```

Comment:
The TestRPCTimeout-4 failure (https://travis-ci.org/grpc/grpc-go/builds/108934236) is before this change. The problem is that recvMsg only expects transport.ConnectionError and transport.StreamError from the transport layer.

With this change, it is not clear to me what kind of errors can be returned by Body.Read so that I am not sure how to handle them correctly.